Re-apply "Mandatory linking pass"

slavapestov opened the pull request:
As a first step to getting mandatory inlining out of the business
of 'linking' (walking the function graph and deserializing all
referenced functions), add a new optimizer pass which links
everything in the mandatory pipeline.

For now this is mostly NFC, except it regresses an optimization
I made recently by linking in bodies of methods of deserialized
vtables eagerly. This will be addressed in upcoming patches.

slavapestov committed:
Mandatory Inlining: Remove outdated remark about transparent functions

Transparent function bodies are in fact available externally.

slavapestov committed:
SILOptimizer: Add 'mandatory SIL linker' pass

As a first step to getting mandatory inlining out of the business
of 'linking' (walking the function graph and deserializing all
referenced functions), add a new optimizer pass which links
everything in the mandatory pipeline.

For now this is mostly NFC, except it regresses an optimization
I made recently by linking in bodies of methods of deserialized
vtables eagerly. This will be addressed in upcoming patches.

slavapestov committed:
Mandatory Inlining: Use loadFunction() instead of linkFunction()

We only need to deserialize the function itself, not its transitive
dependencies. Also, only deserialize a function after we've checked
that its transparent.

For now, this doesn't reduce the volume of SIL linking, because the
mandatory linker pass still links everything. But we're almost
there.

slavapestov committed:
SIL Linker: Deserialize HiddenExternal functions

PublicNonABI function declarations deserialize as HiddenExternal,
then become SharedExternal after the body has been deserialized.
So try deserializing HiddenExternal too.

NFC until mandatory inlining is no longer eagerly deserializing.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@atrick I can't reproduce the failure locally, but it definitely looks like a real issue. Trying PR testing. I won't merge this until I figure out what's going on.

atrick commented:
@slavapestov It was failing consistently on `SDK/Dispatch/watchsimulator/i386/Dispatch.o ` when building with -sil-verify-all. If it passes for you now, then I don't know... maybe the PR testing workspace wasn't clean?

slavapestov committed:
SIL: Don't require shared functions to have bodies in raw SIL

We don't IRGen raw SIL, and once mandatory inlining and mandatory
linking are performed in separate passes, this invariant no longer
holds.

slavapestov commented:
@swift-ci Please test
